### PR TITLE
Backpack: Add find or create channel route

### PR DESCRIPTION
**Pull request by molly-moen**
Add a new route `/backpacks/channel` that returns the encrypted channel token for the current user's backpack, creating the backpack if it does not exist. Any signed in user can create a backpack (which aligns with the permissions on projects).

## Links

- spec: [Backpack](https://docs.google.com/document/d/1Y01X2_rrVt8f4y8yHpm3BBpq31Ra9CSkOLYycwQ5or8/edit#)
- jira ticket: [CSA-522](https://codedotorg.atlassian.net/browse/CSA-522)

## Testing story
Tested locally with new unit tests.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/backpack.rb [hunk 1]
```diff
@@ -13,4 +13,23 @@
...
 class Backpack < ApplicationRecord
+  belongs_to :user
+  validates_uniqueness_of :storage_app_id
+  validates_uniqueness_of :user_id
+
+  def self.find_or_create(user_id, ip)
+    backpack = find_by_user_id(user_id)
+    unless backpack
+      # Create a storage app for this user's backpack
+      storage_app = StorageApps.new(get_storage_id)
```
Comment (bencodeorg) [inline dashboard/app/models/backpack.rb]: Will the record created in `StorageApps` have `user_id` = nil? If so, is that intentional?

**Review by bencodeorg — APPROVED** [review]

File: dashboard/app/models/backpack.rb (hunk 1 above)
Comment (molly-moen) [reply dashboard/app/models/backpack.rb]: storage apps doesn't have a user id field, it instead has `storage_id` (which is unique per user), which is what we are passing in with `get_storage_id`

File: dashboard/app/models/backpack.rb [hunk 2]
```diff
@@ -13,4 +13,23 @@
 #  index_backpacks_on_user_id  (user_id)
 #
 class Backpack < ApplicationRecord
+  belongs_to :user
+  validates_uniqueness_of :storage_app_id
+  validates_uniqueness_of :user_id
```
Comment (bethanyaconnor) [inline dashboard/app/models/backpack.rb]: Is it worth changing the index to be unique?

File: dashboard/app/models/backpack.rb [hunk 3]
```diff
@@ -13,4 +13,23 @@
 #  index_backpacks_on_user_id  (user_id)
 #
 class Backpack < ApplicationRecord
+  belongs_to :user
+  validates_uniqueness_of :storage_app_id
```
Comment (bethanyaconnor) [inline dashboard/app/models/backpack.rb]: Do we need an index on storage_app_id?

File: dashboard/app/models/backpack.rb (hunk 2 above)
Comment (molly-moen) [reply dashboard/app/models/backpack.rb]: I intended to do that [here](https://github.com/code-dot-org/code-dot-org/pull/41635), but looking at the generated schema file it seems I used the wrong syntax. I'll do a follow-up to actually make it unique and remove this here

File: dashboard/app/models/backpack.rb (hunk 3 above)
Comment (molly-moen) [reply dashboard/app/models/backpack.rb]: We will only need to look up by user, not storage_app_id

File: dashboard/app/models/backpack.rb (hunk 1 above)
Comment (bencodeorg) [reply dashboard/app/models/backpack.rb]: Sorry meant the record in `user_storage_ids`, which I think will have a user ID in the actual controller action/test (b/c we have a current_user), and will not have a user ID for the model test you added, or if we ever created backpacks outside the context of a controller (b/c no current_user)? Probably doesn't matter

https://github.com/code-dot-org/code-dot-org/blob/65e1499521235ef7a35053b54a93b903ad108729/shared/middleware/helpers/storage_id.rb#L95

https://github.com/code-dot-org/code-dot-org/blob/65e1499521235ef7a35053b54a93b903ad108729/shared/middleware/helpers/storage_id.rb#L5

File: dashboard/app/models/backpack.rb (hunk 3 above)
Comment (bethanyaconnor) [reply dashboard/app/models/backpack.rb]: I'm not sure but won't Rails query the table for `storage_app_id` when it save a backpack to verify uniqueness?

Comment (molly-moen) [reply dashboard/app/models/backpack.rb]: ah good point. Looking at the [docs](https://guides.rubyonrails.org/active_record_validations.html#uniqueness) for `uniqueness` and it seems that doing this doesn't even always guarantee uniqueness. I'll remove the uniqueness constraints and fix up the indexes as a follow-up.

File: dashboard/app/models/backpack.rb (hunk 1 above)
Comment (molly-moen) [reply dashboard/app/models/backpack.rb]: ah gotcha. I'll update this to explicitly get the storage_id for the given user_id

**Review by bethanyaconnor — APPROVED** [review]

**Review by bencodeorg — APPROVED** [review]

**Pull request merged**